### PR TITLE
add doc to transducer functions

### PR DESCRIPTION
**Pull request by uriva**

### CODE REVIEW

**Comment by uriva** [comment]
automerge

File: gamla/transducer.py [hunk 1]
```diff
@@ -44,22 +70,47 @@ def _replace_index(the_tuple, index, value):
...
+        [],
+        [1, 2, 3]
+    )
+    (3,)
+    """
     return lambda step: lambda s, x: step(s, x) if f(x) else s
 
 
 def concat(step: Reducer):
+    """Flattens collections using the `step` function into a single collection."""
```
Comment (itayzit) [inline gamla/transducer.py:103]: is it the same behaviour as functional_generic.concat? if it is, is there any difference/reason to use transducer? if it isn't, i think an example will be good

File: gamla/transducer.py [hunk 2]
```diff
@@ -44,22 +70,47 @@ def _replace_index(the_tuple, index, value):
...
     return lambda s, x: functools.reduce(step, x, s)
 
 
 def mapcat(f: Callable[[Any], Iterable[Any]]):
+    """Maps over a collection, then flattens the results into a single collection."""
```
Comment (itayzit) [inline gamla/transducer.py:108]: same as concat.

File: gamla/transducer.py [hunk 3]
```diff
@@ -44,22 +70,47 @@ def _replace_index(the_tuple, index, value):
...
     return functional_generic.compose(map(f), concat)
 
 
 def groupby(key: Callable[[Any], Any], reducer: Reducer, initial):
+    """Like `groupby_many`, just with a key function that returns a single element."""
```
Comment (itayzit) [inline gamla/transducer.py:113]: same as concat, mapcat

**Review by itayzit — APPROVED** [review]

File: gamla/transducer.py (hunk 1 above)
Comment (uriva) [reply gamla/transducer.py:103]: yeah it's different, but it's more like an entire talk why it's different, and relevant to all the functions in this namespace. The users of `transducer` are asumed to have this knowledge.

The basic idea is that it doesn't assume anything about the collection (so it's not a generator, for example). One benefit is not having to worry about generator depletion.

wdyt?

**Pull request merged**